streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -141,39 +141,44 @@ private void initialize(final RestoringTasks active) {
 
         // try to fetch end offsets for the initializable restorers and remove any partitions
         // where we already have all of the data
+        final Map<TopicPartition, Long> queriedEndOffsets;
         try {
-            endOffsets.putAll(restoreConsumer.endOffsets(initializable));
+            queriedEndOffsets = restoreConsumer.endOffsets(initializable);
         } catch (final TimeoutException e) {
             // if timeout exception gets thrown we just give up this time and retry in the next run loop
             log.debug("Could not fetch end offset for {}; will fall back to partition by partition fetching", initializable);
             return;
         }
 
+        queriedEndOffsets.forEach((tp, endOffset) -> {
+            if (endOffset != null) {
+                final StateRestorer restorer = stateRestorers.get(tp);
+                final long offsetLimit = restorer.offsetLimit();
+                endOffsets.put(tp, Math.min(endOffset, offsetLimit));
```

Comment:
`endOffsets` is usually used for the topic-partition end-offsets -- however, the new semantics is different -- maybe rename to `restoreOffset` (hope you find a better name) for avoid confusion?

`queriedEndOffsets` can than be `endOffsets` what is accurate and aligns with all other code.

Reply:
Good idea re. re-working the loop!

I agree on overloading the semantics of endOffsets. I hesitated on that one. I'll rename to make it clearer.

Reply:
Kind of funny. With the early out (in `StandbyTask`) it raises the NPath complexity above the threshold: 

```
[ant:checkstyle] [ERROR] /Users/cpettitt/projects/kafka/streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:127:5: NPath Complexity is 504 (max allowed is 500). [NPathComplexity]
```

Is this something we can raise? The new code (your recommendation) is clearly simpler to read. Otherwise I'm going to have to go hunting for some code somewhere to refactor to get under this limit.